### ISSUE TITLE
About data preprocessing.

### ISSUE THREAD

**Issue by KaKa-101** [Chat-3D/Chat-3D-v2#31]
Thanks for your work and when I preprocessed the data guided by readme, I met the following two questions:
1. First, this link seems to point to an empty file.
![image](https://github.com/Chat-3D/Chat-3D-v2/assets/166525403/f52debb7-8bda-4638-999b-07a20c96403e)
And in the [run_prepare.sh](https://github.com/Chat-3D/Chat-3D-v2/blob/main/preprocess/run_prepare.sh), the value of "version" is set to null. Is that right?
![image](https://github.com/Chat-3D/Chat-3D-v2/assets/166525403/f7f7c9a3-a5ad-4143-a1f3-460b65cfb7c7)
Also it seems that there isn't the file `inference.sh` yet and I don't know whether other files in [preprocess](https://github.com/Chat-3D/Chat-3D-v2/tree/main/preprocess) need to be updated or not.
![image](https://github.com/Chat-3D/Chat-3D-v2/assets/166525403/6eea1bf4-cd1e-4827-89d9-80376a2733b4)
3. Second, when I prepare the conda environment as said in [mask3d](https://github.com/JonasSchult/Mask3D?tab=readme-ov-file), the following error appears:
![image](https://github.com/Chat-3D/Chat-3D-v2/assets/166525403/c6cd69ee-998f-45b8-9a9e-5f2c470e9b12)
It seems that "opencv==0.17.0"(or other versions of opencv) cannot be installed  in the setting of python-3.10.9. Have you ever encountered this problem and how should I fix it ?
Appreciate for your valuable work and help.

**Comment by ZzZZCHS** [comment]
1. The `run_prepare.sh` refers to `preprocess/run_prepare.sh`. Thanks for pointing out the link problem. 
The `version` is set to null as default. It's right.
You can find `scripts/inference.sh` in this [repo](https://github.com/ZzZZCHS/Uni3D) which was forked from Uni3D.
2. It's a little tricky to install open3d package. This package is only used in their visualization code. So we skipped installing this package and directly run the inference code.

**Comment by KaKa-101** [comment]
> 1. The `run_prepare.sh` refers to `preprocess/run_prepare.sh`. Thanks for pointing out the link problem.
>    The `version` is set to null as default. It's right.
>    You can find `scripts/inference.sh` in this [repo](https://github.com/ZzZZCHS/Uni3D) which was forked from Uni3D.
> 2. It's a little tricky to install open3d package. This package is only used in their visualization code. So we skipped installing this package and directly run the inference code.

Thanks for your reply. Could you tell me what's the destination of the step `Process data` after getting every instance segmentation of  each scene using pretrained Mask3D model? 
![image](https://github.com/Chat-3D/Chat-3D-v2/assets/166525403/032c7576-9f99-4356-b1a7-02b651180d54)
And after conducting `bash preprocess/run_prepare.sh` using `mask3d_inst_seg` you have provided, I met the following error:
![image](https://github.com/Chat-3D/Chat-3D-v2/assets/166525403/1da4b9a9-7bf1-4171-85df-42eb8d5afba3)

**Comment by ZzZZCHS** [comment]
1. The destination of the step `Process data` is prepare QA pairs for each task/dataset, and transform the GT IDs to correponding segmented IDs (based on IoU between GT instances and segmented instances)
2. This is my problem. Some of them are stilling requiring some old files which I haven't provided. I will fix these code soon (to only rely on the official annotations of each dataset)

**Comment by ZzZZCHS** [comment]
Hi @KaKa-101, we've updated the preprocessing codes to only rely on official annotations of each dataset.

**Comment by KaKa-101** [comment]
Thanks for your great  work. But when I prepare the environment for [uni3d](https://github.com/ZzZZCHS/Uni3D), when I run `pip install "git+git://github.com/erikwijmans/Pointnet2_PyTorch.git#egg=pointnet2_ops&subdirectory=pointnet2_ops_lib"
`, I got the following error:
![468e77df2d1a25e01b1873841808263](https://github.com/Chat-3D/Chat-3D-v2/assets/166525403/2ac99ed4-cec4-4e58-9316-09bef85057d1)
Do you know how to fix it? I'd appreciate it very much.

**Comment by ZzZZCHS** [comment]
I just now followed the instruction of uni3d to install the environment and got the same error as yours.

Then I find the problem is related to the pytorch version. You would install pytorch 2.3.0 using his scripts, which is probably uncompatible with pointnet2_ops for now. I switch it to pytorch 2.0.1 (which is the version we used before), then it works.

I remmond you trying this:
```
conda install pytorch==2.0.1 torchvision==0.15.2 torchaudio==2.0.2 pytorch-cuda=11.8 -c pytorch -c nvidia -y
pip install "git+https://github.com/erikwijmans/Pointnet2_PyTorch.git#egg=pointnet2_ops&subdirectory=pointnet2_ops_lib"
```

**Comment by KaKa-101** [comment]
Thanks for your advice and it works.
But I don't get the desired `segment_result_dir` using pretrained Mask3D to do instance segmentation. First, I preprocess the datasets like this (--scannet200 is setted to False) guided in [Mask3D](https://github.com/JonasSchult/Mask3D/tree/main?tab=readme-ov-file#data-preprocessing-hammer):
![image](https://github.com/Chat-3D/Chat-3D-v2/assets/166525403/b30330a8-576f-44ec-9e47-ec1e606ba3c4)
Then I conduct validation in ScanNet like this:
![image](https://github.com/Chat-3D/Chat-3D-v2/assets/166525403/21639866-9d47-463c-b160-d409d461e611)
But in the `saved` folder, there is nothing.
![image](https://github.com/Chat-3D/Chat-3D-v2/assets/166525403/ca2751f0-420a-4870-b879-9bc6c4d6965b)
So do you know how should I change the settings or anything? Besides I want to know why you choose `scannet200_val.pt` but not `scannet.pt` as pretrained checkpoint since we are conducting experiments on ScanNet dataset. Thanks very much for your help and I'd appreciate it a lot.

**Comment by ZzZZCHS** [comment]
This is the script I used:
```shell
#!/bin/bash
export OMP_NUM_THREADS=3  # speeds up MinkowskiEngine

NUM_GPUS=1
CURR_DBSCAN=0.95
CURR_TOPK=750
CURR_QUERY=100

# TEST
python main_instance_segmentation.py \
general.gpus=${NUM_GPUS} \
general.experiment_name="test0416_scannet200" \
general.project_name="test0416" \
general.checkpoint='checkpoints/scannet200/scannet200_val.ckpt' \
data/datasets=scannet200 \
general.num_targets=201 \
data.num_labels=200 \
general.train_mode=false \
general.eval_on_segments=true \
general.train_on_segments=true \
model.num_queries=${CURR_QUERY} \
general.topk_per_image=${CURR_TOPK} \
general.use_dbscan=true \
general.dbscan_eps=${CURR_DBSCAN} \
general.save_visualizations=false \
general.export=true
```

Basically, you need to set `general.export=true` to let the model export predicted masks. The `eval_output` folder will look like this:
![Screenshot 2024-05-07 at 10 15 01 PM](https://github.com/Chat-3D/Chat-3D-v2/assets/50910784/42657c2d-602c-43c7-ba5b-2d3cc910013e)

And with the default setting, you can only get the result of val split. To get the result of train split, I directly changed this [line](https://github.com/JonasSchult/Mask3D/blob/main/datasets/semseg.py#L193) to `self._load_yaml(database_path / f"train_database.yaml")`


ScanNet and ScanNet200 share the same scene data. The only difference between them is the category annotation (the latter one has 200 classes). In our experiment, we find that the model trained on ScanNet200 generates a better instance segmentation result. Plus we do not need the prediceted class labels for our model, so using `scannet200_val.pt` should be ok.

**Comment by KaKa-101** [comment]
Thanks for your quick reply, which helps me a lot. 
And I want to know if I use `scannet200_val.pt` as pretrained_weight, do I need to set `--scannet200` to True in the step of preprocessing dataset:
![image](https://github.com/Chat-3D/Chat-3D-v2/assets/166525403/09318665-6f06-4dc8-96fe-d4b9b07c9f53)

**Comment by ZzZZCHS** [comment]
Yes, you need to set `--scannet200=true`.

**Comment by KaKa-101** [comment]
Thanks a lot.
Besides, in the paper you mention that you designed a relation module to incorporate spatial information into a relation-aware token for each object in the scene-level alignment. Could you explain this part in detail and which part of the code you provided corresponds to the `relation module`?

**Comment by ZzZZCHS** [comment]
We proposed the relation module to get a scene-aware token for each object [here](https://github.com/Chat-3D/Chat-3D-v2/blob/main/models/chat3d.py#L319-L328).

However, we've discarded it in the released version of v2.1 since we find these scene-aware tokens do not help improve the performance. We are still exploring how to incorporate position information into the model in a better way.

**Comment by KaKa-101** [comment]
Thanks for your reply.
And could you tell me what's the difference between `scannet_train/val_attributes.pt` and `scannet_mask3d_train/val_attributes.pt`? It seems that both of them are consist of labels and points' coordinates of all instances in each scene, the former is getted from official annotations in ScanNet dataset and the latter is getted from the instance segmentation results using pretrained Mask3D?  And what's the function of these files in subsequent operations?

**Comment by ZzZZCHS** [comment]
2. Yes. `scannet_train/val_attributes.pt` saves the location (3D center and size, which can be transformed to 3D bounding box) and class label of each GT instance. While `scannet_mask3d_train/val_attributes.pt` saves the location and class label of each segmented instance (from Mask3D).
3. `scannet_train/val_attributes.pt` is only used for evaluation (calculate the IoU between GT instance's bbox and segmented instance's bbox). We only input the segmented instances to the model (both for training and evaluating).

**Comment by KaKa-101** [comment]
Thanks for your reply, which helps a lot.
And I find that there is a file named `scan2cap_val_corpus.json` in your provided [annotations](https://drive.google.com/drive/folders/1feFAsDmeDmx_PVLcuQm8aZFkNuUGz4Gu), but it seemed I can't generate this file using provided preprocessing code. So what's the function of this file and do I need it for subsequent training and inference?
![image](https://github.com/Chat-3D/Chat-3D-v2/assets/166525403/c1bac208-60e8-4cf7-9348-cae317b69117)

**Comment by ZzZZCHS** [comment]
It's used in Scan2Cap evaluation. We reference Scan2Cap's [code](https://github.com/daveredrum/Scan2Cap/blob/main/lib/eval_helper.py#L33-L54) to prepare this corpus. Basically, it converts all the original descriptions into "sos {description} eos".

**Comment by KaKa-101** [comment]
> This is the script I used:
> 
> ```shell
> #!/bin/bash
> export OMP_NUM_THREADS=3  # speeds up MinkowskiEngine
> 
> NUM_GPUS=1
> CURR_DBSCAN=0.95
> CURR_TOPK=750
> CURR_QUERY=100
> 
> # TEST
> python main_instance_segmentation.py \
> general.gpus=${NUM_GPUS} \
> general.experiment_name="test0416_scannet200" \
> general.project_name="test0416" \
> general.checkpoint='checkpoints/scannet200/scannet200_val.ckpt' \
> data/datasets=scannet200 \
> general.num_targets=201 \
> data.num_labels=200 \
> general.train_mode=false \
> general.eval_on_segments=true \
> general.train_on_segments=true \
> model.num_queries=${CURR_QUERY} \
> general.topk_per_image=${CURR_TOPK} \
> general.use_dbscan=true \
> general.dbscan_eps=${CURR_DBSCAN} \
> general.save_visualizations=false \
> general.export=true
> ```
> 
> Basically, you need to set `general.export=true` to let the model export predicted masks. The `eval_output` folder will look like this: ![Screenshot 2024-05-07 at 10 15 01 PM](https://private-user-images.githubusercontent.com/50910784/328557221-42657c2d-602c-43c7-ba5b-2d3cc910013e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYzNjg0MjksIm5iZiI6MTcxNjM2ODEyOSwicGF0aCI6Ii81MDkxMDc4NC8zMjg1NTcyMjEtNDI2NTdjMmQtNjAyYy00M2M3LWJhNWItMmQzY2M5MTAwMTNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIyVDA4NTUyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZkZjgwOTgxZmYyNzRhY2EyMjk4ZDYyYTk0ZTQzNzkzNDkyN2YzOTRlNWM4ZTI4NTM3MjY1ZTcyOGVkNWNiNzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.sFkPeTrYdjgVF0qhC4N1UK606CT50ex7PEWPeYErsyk)
> 
> And with the default setting, you can only get the result of val split. To get the result of train split, I directly changed this [line](https://github.com/JonasSchult/Mask3D/blob/main/datasets/semseg.py#L193) to `self._load_yaml(database_path / f"train_database.yaml")`
> 
> ScanNet and ScanNet200 share the same scene data. The only difference between them is the category annotation (the latter one has 200 classes). In our experiment, we find that the model trained on ScanNet200 generates a better instance segmentation result. Plus we do not need the prediceted class labels for our model, so using `scannet200_val.pt` should be ok.

Thanks for your provided scipt of Mask3D. Do you know in which file  I can learn or configure all the parameters. For example, I want to know where the setting `general.export=true` takes effect. I haven't figured out how these parameters work in the code of Mask3D. 
I will appreciate it a lot if you can provide help.

**Comment by ZzZZCHS** [comment]
It takes effect [here](https://github.com/JonasSchult/Mask3D/blob/main/trainer/trainer.py#L964).
The basic configs are in this [file](https://github.com/JonasSchult/Mask3D/blob/main/conf/config_base_instance_segmentation.yaml).
Their code is based on pytorch lightning, maybe you need to find some tutorials about this.
